src/set_theory/cardinal/finite.lean

```diff
@@ -120,4 +120,168 @@ lemma card_eq_coe_fintype_card [fintype α] : card α = fintype.card α := mk_to
 @[simp]
 lemma card_eq_top_of_infinite [infinite α] : card α = ⊤ := mk_to_part_enat_of_infinite
 
+
+lemma card_congr {α : Type*} {β : Type*} (f : α ≃ β) :
+  part_enat.card α = part_enat.card β :=
+cardinal.to_part_enat_congr f
+
+lemma card_ulift (α : Type*) : card (ulift α) = card α :=
+card_congr equiv.ulift
+
+@[simp] lemma card_plift (α : Type*) : card (plift α) = card α :=
+card_congr equiv.plift
+
+lemma card_image_of_inj_on {α : Type*} {β : Type*} {f : α → β} {s : set α} (h : set.inj_on f s) :
+  card (f '' s) = card s :=
+card_congr (equiv.set.image_of_inj_on f s h).symm
+
+lemma card_image_of_injective {α : Type*} {β : Type*}
+  (f : α → β) (s : set α) (h : function.injective f) :
+  card (f '' s) = card s :=
+card_congr (equiv.set.image_of_inj_on f s (set.inj_on_of_injective h s)).symm
+
+lemma succ_le_iff {n : ℕ} {e : part_enat} : ↑n.succ ≤ e ↔ ↑n < e:=
+begin
+  rw [coe_lt_iff, coe_le_iff],
+  apply forall_congr, intro a, rw nat.succ_le_iff,
+end
+
+/- The line
+-- obtain ⟨m, hm⟩ := cardinal.lt_aleph_0.mp h,
+extract an integer m from a cardinal c such that h : c < ℵ₀
+It may appear easier to use than the rewrites I finally use … -/
+
+lemma coe_nat_le_iff_le {n : ℕ} {c : cardinal} :
+   ↑n ≤ to_part_enat c ↔ ↑n ≤ c :=
+begin
+  cases lt_or_ge c aleph_0,
+  { rw [to_part_enat_apply_of_lt_aleph_0 h, coe_le_coe, ← to_nat_cast n],
+    rw to_nat_le_iff_le_of_lt_aleph_0 (nat_lt_aleph_0 n) h,
+    simp only [to_nat_cast] },
+  { apply iff_of_true,
+    { rw to_part_enat_apply_of_aleph_0_le h,
+      exact le_top },
+    { apply le_trans (le_of_lt _) h,
+      rw lt_aleph_0,
+      use n } }
+end
+
+lemma le_coe_nat_iff_le {c : cardinal} {n : ℕ} :
+   to_part_enat c ≤ n ↔ c ≤ n :=
+begin
+  cases lt_or_ge c aleph_0,
+  { rw [to_part_enat_apply_of_lt_aleph_0 h, coe_le_coe, ← to_nat_cast n],
+    rw to_nat_le_iff_le_of_lt_aleph_0 h (nat_lt_aleph_0 n),
+    simp only [to_nat_cast] },
+  { apply iff_of_false,
+    { rw to_part_enat_apply_of_aleph_0_le h,
+      simp only [top_le_iff, coe_ne_top, not_false_iff] },
+    { rw not_le,
+      apply lt_of_lt_of_le (nat_lt_aleph_0 n) h } }
+end
+
+lemma coe_nat_eq_iff_eq {n : ℕ} {c : cardinal} :
+   ↑n = to_part_enat c ↔ ↑n = c :=
+begin
+  cases lt_or_ge c aleph_0,
+  { rw [to_part_enat_apply_of_lt_aleph_0 h, coe_inj],
+    rw ← to_nat_cast n,
+    rw to_nat_eq_iff_eq_of_lt_aleph_0 (nat_lt_aleph_0 n) h,
+    simp only [to_nat_cast] },
+  { apply iff_of_false,
+    { rw to_part_enat_apply_of_aleph_0_le h,
+      exact coe_ne_top n },
+    { apply ne_of_lt,
+      apply lt_of_lt_of_le (nat_lt_aleph_0 n) h } }
+end
+
+lemma eq_coe_nat_iff_eq {c : cardinal} {n : ℕ} :
+   to_part_enat c = n ↔ c = n:=
```

Comment:
`@[simp] lemma cardinal.to_part_enat_eq_coe_nat_iff`, the proof is `by rw [eq_comm, coe_nat_eq_to_part_enat, eq_comm]`

Reply:
Done